docs/Training-On-Your-Data.md

```diff
@@ -0,0 +1,354 @@
+
+If you want a different model size or architecture, you can look at the config files in
+[levanter/config](https://github.com/stanford-crfm/levanter/tree/main/config).
+
+### Continued Pretraining
+
+Levanter supports starting from an HF pretrained model. To do so, you should set your config like this:
+
+```yaml
+model:
+  type: mpt
+initialize_from_hf: "mosaicml/mpt-7b@68e1a8e0ebb9b30f3c45c1ef6195980f29063ae2"
+use_hf_model_config: true
+```
+
+TODO: add instructions for llama including auth.
+
+You should probably reduce the learning rate by a factor of 10 or so. TODO: figure out best practices here.
+
+
+### Checkpointing
+
+See also the [Checkpointing section of the Configuration Guide](./Configuration-Guide.md#checkpointing).
+
+Levanter supports checkpointing to both local and Google Cloud Storage, backed by [Tensorstore](https://google.github.io/tensorstore/).
+If you're using multiple machines, you should probably use cloud storage or NFS.
+
+Levanter saves two kinds of checkpoints:
+
+* **time-based checkpoints**: **temporary** checkpoints that are saved every `save_interval` minutes. The previous
+time-based checkpoint is deleted when a new one is saved.
+* **step-based checkpoints**: **permanent** checkpoints that are saved according to a policy. These checkpoints are never deleted.
+
+At a minimum, you should set a `base_path` for your checkpoints. This can be a local path or a GCS path.
+
+### Determining Batch Size
+
+We don't have a mechanism for tuning batch size automatically, so you need to do this by hand. Memory usage is controlled
+by `per_device_parallelism`, `train_batch_size`, gradient checkpointing, model size, and number of accelerators, and
+`model_axis_size`. (For most models `model_axis_size` should be 1, so you can ignore it.)
+
+`per_device_parallelism` is analogous to `per_device_batch_size` in other frameworks. It controls how many examples
+are processed at once on a single accelerator, but the name is a bit more "correct" in the presence of tensor or pipeline
+parallelism.  A `per_device_parallelism` of `-1` means "use as many as possible to not perform gradient accumulation."
+
+Gradient accumulation is performed whenever `num_accelerators * per_device_parallelism / model_axis_size < train_batch_size`.
+Gradient checkpointing is enabled by default and highly recommended.
+
+So, to find your batch size, you should modify either `per_device_parallelism` or `train_batch_size` until
+your job runs. Note that, due to FSDP, as you add more TPUs, you can increase the effective parallelism because
+you will use less memory per accelerator to store parameters and optimizer states.
+
+### Number of Training Steps
+
+Levanter does not support epochs or number of tokens/examples, so if you want to train for a certain number of epochs or
+tokens, you'll need to compute the number of steps yourself. You can use the following formula:
+
+```
+num_train_steps = num_epochs * num_tokens_per_epoch / train_batch_size / seq_len
```

Comment:
If # tokens in an example is larger than seq_len, do we break it into multiple examples?

Reply:
we do